packages/tally-ho/src/index.ts

```diff
@@ -0,0 +1,158 @@
+   * method is used in an useEffect then it results in 2 requests.
+   */
+  isActivationInProgress = false
+
+  /**
+   * @param connectEagerly - A flag indicating whether connection should be initiated when the class is constructed.
+   */
+  constructor(actions: Actions, connectEagerly?: boolean | ConnectEagerlyConfig) {
+    super(actions)
+
+    if (connectEagerly) {
+      if (connectEagerly === true) {
+        void this.connectEagerly()
+      } else if (isConnectEagerlyConfig(connectEagerly)) {
+        void this.connectEagerly(connectEagerly)
+      } else {
+        throw new Error(
+          `connectEagerly is expected to be 'true' OR {retries: number, timeoutMs: number}, but ${JSON.stringify(
+            connectEagerly
+          )} was received instead.`
+        )
+      }
+    }
+  }
+
+  /** {@inheritdoc Connector.connectEagerly} */
+  public async connectEagerly(config: ConnectEagerlyConfig = { retries: 5, timeoutMs: 500 }): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      let counter = 0
+
+      /**
+       * When `connectEagerly` is set to true at class initialization time there is no window present.
+       * But it is usually ready in 1-2 seconds, so we shall wait for it.
+       */
+      function waitForWindow() {
+        if (typeof window === 'undefined') {
+          counter++
+
+          if (counter === config.retries) {
+            reject(
+              `window was not present after ${config.retries} reties (with ${config.timeoutMs}ms wait period). Try using the connectEagerly method in a useEffect`
+            )
+          }
+
+          setTimeout(waitForWindow, config.timeoutMs)
+        } else {
+          resolve()
+        }
+      }
+
+      waitForWindow()
+    }).then(() => this.activate())
+  }
+
+  /** {@inheritdoc Connector.activate} */
+  public async activate(): Promise<void> {
+    if (window === undefined) {
+      throw new Error("window is not defined. Toto, I have a feeling we're not in Kansas anymore! 🌪")
+    }
+
+    if (!this.provider) {
+      this.initializeProvider()
+    }
+
+    if (this.isActivationInProgress) return
+
+    this.isActivationInProgress = true
+
+    if (isTally(this.provider)) {
+      const cancelActivation = this.actions.startActivation()
+
+      return this.provider
+        .request({ method: 'eth_requestAccounts' })
+        .then(() =>
+          Promise.all([
+            this.provider?.request({ method: 'eth_chainId' }) as Promise<string>,
+            this.provider?.request({ method: 'eth_accounts' }) as Promise<string[]>, // this is not strictly necessary
+          ]).then(([chainId, accounts]) => {
+            this.actions.update({ chainId: parseChainId(chainId), accounts })
+          })
+        )
+        .catch((error: Error) => {
+          this.actions.reportError(error)
+          cancelActivation()
+        })
+        .finally(() => (this.isActivationInProgress = false))
```

Comment:
We might want to avoid returning assignment here as the intent of return-assignment syntax can sometimes be unclear to someone unfamiliar with the codebase